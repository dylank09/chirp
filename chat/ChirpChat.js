import React, { useState, useRef } from "react";
import { StyleSheet, ScrollView, Text, View } from "react-native";
import { AntDesign, SimpleLineIcons } from "@expo/vector-icons";

import { theme } from "../assets/Theme";
import Message from "./Message";
import SendText from "./SendText";
import LoadingScreen from "../components/LoadingScreen";
import ChatOptions from "./ChatOptions";

import firebase from "firebase/app";
import "firebase/firestore";
import {
  useCollectionData,
  useDocumentData,
} from "react-firebase-hooks/firestore";
import app from "../config/FirebaseConfig";

const firestore = firebase.firestore(app);
const auth = firebase.auth();

export default function ChirpChat({ name, id, onBackPress }) {
  const [optionsOpen, setOptionsOpen] = useState(false);

  const scrollViewRef = useRef();

  const { uid } = auth.currentUser;
  const userRef = firestore.collection("users").doc(uid);
  const [user] = useDocumentData(userRef);

  const chatRef = firestore.collection("chatGroups").doc(id);
  const [chat] = useDocumentData(chatRef);

  const messagesRef = chatRef.collection("messages");
  const query = messagesRef.orderBy("createdAt").limit(50);
  const [msgs, loading] = useCollectionData(query, { idField: "msgId" });

  async function sendText(text) {
    await messagesRef.add({
      text: text ? text : "",
      createdAt: firebase.firestore.FieldValue.serverTimestamp(),
      uid,
      user: user.name,
    });

    firestore
      .collection("chatGroups")
      .doc(id)
      .update({
        membersUnseen: chat.members,
        lastMessage: text.length < 30 ? text : text.slice(0, 30) + "...",
        lastMessageTimestamp: firebase.firestore.FieldValue.serverTimestamp(),
      });
  }

  function openOptions() {
    setOptionsOpen(true);
  }

  if (loading) {
    return <LoadingScreen />;
  } else if (optionsOpen) {
    return (
      <ChatOptions
        name={name}
        id={id}
        chatData={chat}
        returnToChat={() => setOptionsOpen(false)}
      ></ChatOptions>
    );
  } else {
    return (
      <View style={styles.container}>
        <View style={styles.header}>
          <AntDesign
            style={styles.back}
            name="left"
            size={24}
            color="white"
            onPress={onBackPress}
          />
          <Text style={styles.chatName}>{name}</Text>
          <SimpleLineIcons
            style={styles.options}
            name="options-vertical"
            size={22}
            color="white"
            onPress={openOptions}
          />
        </View>
        <ScrollView
          style={styles.messagesBox}
          ref={scrollViewRef}
          onContentSizeChange={() =>
            scrollViewRef.current.scrollToEnd({ animated: false })
          }
        >
          {msgs &&
            msgs.map((msg) => (
              <Message
                key={msg.msgId}
                text={msg.text}
                timestamp={msg.createdAt ? msg.createdAt.seconds : "  "}
                me={msg.uid == uid}
                user={msg.user}
              />
            ))}
        </ScrollView>
        <View style={styles.bottomBar}>
          <SendText send={sendText}></SendText>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "column",
    width: "100%",
    height: "100%",
    backgroundColor: theme.colors.background,
    alignItems: "center",
    justifyContent: "space-between",
  },
  header: {
    flexDirection: "row",
    width: "100%",
    justifyContent: "space-between",
    marginVertical: 15,
  },
  messagesBox: {
    flex: 1,
    width: "100%",
  },
  bottomBar: {
    width: "100%",
    justifyContent: "center",
  },
  back: {
    marginLeft: 10,
  },
  options: {
    alignSelf: "flex-end",
    marginRight: 10,
  },
  chatName: {
    color: theme.colors.text,
    width: "100%",
    textAlign: "center",
    marginRight: 25,
    fontWeight: "500",
    fontSize: theme.dimensions.standardFontSize,
  },
});
