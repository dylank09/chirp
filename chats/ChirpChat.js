import React, { useState, useRef } from "react";
import {
  StyleSheet,
  ScrollView,
  Text,
  View,
  KeyboardAvoidingView,
} from "react-native";
import { AntDesign, SimpleLineIcons } from "@expo/vector-icons";

import { theme } from "../assets/Theme";
import Message from "./Message";
import SendText from "./SendText";
import LoadingScreen from "../components/LoadingScreen";
import ChatOptions from "./ChatOptions";

import {
  useCollectionData,
  useDocumentData,
} from "react-firebase-hooks/firestore";

import firebase from "../config/FirebaseInit";
import firestore from "../config/FirestoreInit";
import auth from "../config/FirebaseAuthInit";

export default function ChirpChat({ name, id, onBackPress }) {
  const [optionsOpen, setOptionsOpen] = useState(false);
  const [text, setText] = useState("");

  const scrollViewRef = useRef();

  const currentUser = auth.currentUser;
  var uid, email;
  if (currentUser) {
    uid = currentUser.uid;
    email = currentUser.email;
  }
  const userRef = firestore.collection("users").doc(uid);
  const [user] = useDocumentData(userRef);

  const chatRef = firestore.collection("chatGroups").doc(id);
  const [chat] = useDocumentData(chatRef);

  const messagesRef = chatRef.collection("messages");
  const query = messagesRef.orderBy("createdAt").limit(50);
  const [msgs, loading] = useCollectionData(query, { idField: "msgId" });

  async function sendText() {
    if (text.length > 0) {
      await messagesRef.add({
        text: text ? text : "",
        createdAt: firebase.firestore.FieldValue.serverTimestamp(),
        uid,
        email,
        user: user.name,
      });

      firestore
        .collection("chatGroups")
        .doc(id)
        .update({
          membersUnseen: chat.members,
          lastMessage: text.length < 30 ? text : text.slice(0, 30) + "...",
          lastMessageTimestamp: firebase.firestore.FieldValue.serverTimestamp(),
        });

      setText("");
    }
  }

  function openOptions() {
    setOptionsOpen(true);
  }

  if (loading) {
    return <LoadingScreen />;
  } else if (optionsOpen) {
    return (
      <ChatOptions
        name={name}
        id={id}
        chatData={chat}
        returnToChat={() => setOptionsOpen(false)}
        returnToMain={onBackPress}
      ></ChatOptions>
    );
  } else {
    return (
      <View style={styles.container}>
        <View style={styles.header}>
          <AntDesign
            name="left"
            size={24}
            color="white"
            testID="backButton"
            onPress={onBackPress}
          />
          <Text style={styles.chatName} testID="chatName">
            {name}
          </Text>
          <SimpleLineIcons
            name="options-vertical"
            size={22}
            color="white"
            onPress={openOptions}
          />
        </View>
        <ScrollView
          style={styles.messagesBox}
          ref={scrollViewRef}
          onContentSizeChange={() =>
            scrollViewRef.current.scrollToEnd({ animated: true })
          }
        >
          {msgs && msgs.length > 0 ? (
            msgs.map((msg) => (
              <Message
                key={msg.msgId}
                text={msg.text}
                timestamp={msg.createdAt ? msg.createdAt.seconds : "  "}
                me={msg.uid == uid}
                user={msg.user}
              />
            ))
          ) : (
            <Text style={styles.noMessagesText} testID="noChatsHelpText">
              Be the first to send a message! {"\n\n\n\n"} Click on the â‹® in the
              top right corner to open options and add other members
            </Text>
          )}
        </ScrollView>
        <KeyboardAvoidingView
          style={styles.bottomBar}
          behavior="padding"
          keyboardVerticalOffset={-150}
        >
          <SendText
            text={text}
            setText={setText}
            send={sendText}
            testID="sendTextBox"
          ></SendText>
        </KeyboardAvoidingView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "column",
    width: "100%",
    height: "100%",
    backgroundColor: theme.colors.background,
    alignItems: "center",
    justifyContent: "space-between",
  },
  header: {
    flexDirection: "row",
    width: "90%",
    justifyContent: "space-between",
    marginTop: 20,
    marginBottom: 20,
  },
  messagesBox: {
    flex: 1,
    width: "100%",
  },
  bottomBar: {
    width: "100%",
    justifyContent: "center",
  },
  chatName: {
    color: theme.colors.text,
    textAlign: "center",
    fontWeight: "bold",
    fontSize: theme.dimensions.standardFontSize + 2,
  },
  noMessagesText: {
    textAlign: "center",
    width: "80%",
    alignSelf: "center",
    color: theme.colors.hazeText,
    fontSize: theme.dimensions.standardFontSize + 2,
  },
});
