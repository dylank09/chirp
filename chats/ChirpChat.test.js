import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react-native";

import ChirpChat from "./ChirpChat";

describe("<ChirpChat />", () => {
  const tree = renderer.create(<ChirpChat />).toJSON();

  it("renders correctly", () => {
    expect(tree).toMatchSnapshot();
  });

  it("loading screen is present", () => {
    const { getByTestId } = render(<ChirpChat />);
    const foundElement = getByTestId("loading");
    expect(foundElement).toBeTruthy();
  });

  //   it("help text is present", () => {
  //     const { getByTestId } = render(<ChirpChat />);
  //     const foundText = getByTestId("noChatsHelpText");
  //     expect(foundText).toBeTruthy();
  //   });

  //   it("back button is present", () => {
  //     const { getByTestId } = render(<ChirpChat />);
  //     const foundButton = getByTestId("backButton");
  //     expect(foundButton).toBeTruthy();
  //   });

  //   it("chat name is present", () => {
  //     const { getByTestId } = render(<ChirpChat />);
  //     const foundButton = getByTestId("chatName");
  //     expect(foundButton).toBeTruthy();
  //   });

  //   it("send text box is present", () => {
  //     const { getByTestId } = render(<ChirpChat />);
  //     const foundElement = getByTestId("sendTextBox");
  //     expect(foundElement).toBeTruthy();
  //   });
});
