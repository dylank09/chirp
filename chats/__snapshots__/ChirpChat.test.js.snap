// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChirpChat /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(76, 76, 71)",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="loading"
>
  <Image
    source={1}
    style={
      Object {
        "height": 120,
        "width": 112,
      }
    }
  />
</View>
`;
